python/dask_cudf/dask_cudf/batcher_sortnet.py

```diff
@@ -137,3 +146,176 @@ def sort_delayed_frame(parts, by):
     valid = compute(valid_ct)[0]
     validparts = parts[:valid]
     return validparts
+
+
+def _concat(args):
+    if not args:
+        return args
+    args2 = [i for i in args if len(i)]
+    return (
+        args[0]
+        if not args2
+        else gd.concat(args2, axis=0, ignore_index=True, sort=None)
+    )
+
+
+def set_partitions_hash(df, columns, npartitions):
+    c = hash_object_dispatch(df[columns], index=False)
+    return np.mod(c, npartitions)
+
+
+def _shuffle_group_2(df, columns):
+    if not len(df):
+        return {}, df
+    ind = hash_object_dispatch(df[columns], index=False)
+    n = ind.max() + 1
+    result2 = dict(
+        zip(
+            range(n),
+            df.scatter_by_map(
+                ind.values.view(np.int64), map_size=n, keep_index=False
+            ),
+        )
+    )
+    return result2, df.iloc[:0]
+
+
+def _shuffle_group_get(g_head, i):
+    g, head = g_head
+    if i in g:
+        return g[i]
+    else:
+        return head
+
+
+def _shuffle_group(df, columns, stage, k, npartitions):
+    c = hash_object_dispatch(df[columns], index=False)
+    typ = np.min_scalar_type(npartitions * 2)
+    c = np.mod(c, npartitions).astype(typ, copy=False)
+    np.floor_divide(c, k ** stage, out=c)
+    np.mod(c, k, out=c)
+    return dict(
+        zip(
+            range(k),
+            df.scatter_by_map(
+                c.astype(np.int64), map_size=k, keep_index=False
```

Comment:
I might be completely off but does this need to be an `int64`. Feel free to ignore me if i am wrong.

 the reason i am asking is this showed up in a `oom` stack trace  while doing this  for a workflow. 



```python
    return result.compute()
  File "/home/dev/conda/envs/rapids/lib/python3.7/site-packages/dask/base.py", line 166, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "/home/dev/conda/envs/rapids/lib/python3.7/site-packages/dask/base.py", line 437, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/home/dev/conda/envs/rapids/lib/python3.7/site-packages/distributed/client.py", line 2595, in get
    results = self.gather(packed, asynchronous=asynchronous, direct=direct)
  File "/home/dev/conda/envs/rapids/lib/python3.7/site-packages/distributed/client.py", line 1893, in gather
    asynchronous=asynchronous,
  File "/home/dev/conda/envs/rapids/lib/python3.7/site-packages/distributed/client.py", line 780, in sync
    self.loop, func, *args, callback_timeout=callback_timeout, **kwargs
  File "/home/dev/conda/envs/rapids/lib/python3.7/site-packages/distributed/utils.py", line 348, in sync
    raise exc.with_traceback(tb)
  File "/home/dev/conda/envs/rapids/lib/python3.7/site-packages/distributed/utils.py", line 332, in f
    result[0] = yield future
  File "/home/dev/conda/envs/rapids/lib/python3.7/site-packages/tornado/gen.py", line 735, in run
    value = future.result()
  File "/home/dev/conda/envs/rapids/lib/python3.7/site-packages/distributed/client.py", line 1752, in _gather
    raise exception.with_traceback(traceback)
  File "/home/dev/conda/envs/rapids/lib/python3.7/site-packages/dask_cudf-0.13.0a0+3142.gbe2101e56-py3.7.egg/dask_cudf/batcher_sortnet.py", line 201, in _shuffle_group
    c.astype(np.int64), map_size=k, keep_index=False
  File "/home/dev/conda/envs/rapids/lib/python3.7/site-packages/cudf/core/dataframe.py", line 3979, in scatter_by_map
    index_names=index_names,
  File "cudf/_lib/copying.pyx", line 293, in cudf._lib.copying.scatter_to_frames
RuntimeError: RMM error encountered at: /conda/conda-bld/libcudf_1583344320054/work/cpp/src/copying/legacy/copy.cpp:86: 4 RMM_ERROR_OUT_OF_MEMORY
```

Reply:
No - I don't think it does.  This is recycling the logic used in upstream dask - which I suspect is more conservative than necessary.